packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m

```diff
@@ -16,35 +15,26 @@
 
 #import <UMCore/UMEventEmitterService.h>
 
-NSString * const EXDownloadProgressEventName = @"Exponent.downloadProgress";
+#import <EXFileSystem/EXSessionDownloadTaskDelegate.h>
+#import <EXFileSystem/EXSessionUploadTaskDelegate.h>
 
-@interface EXDownloadResumable : NSObject
+NSString * const EXDownloadProgressEventName = @"expo-file-system.downloadProgress";
 
-@property (nonatomic, strong) NSString *uuid;
-@property (nonatomic, strong) NSURLSession *session;
-@property (nonatomic, strong) EXDownloadDelegate *delegate;
+typedef NS_ENUM(NSInteger, EXFileSystemSessionType) {
+  EXFileSystemBackgroundSession = 0,
+  EXFileSystemForegroundSession = 1,
+};
 
-@end
-
-@implementation EXDownloadResumable
-
-- (instancetype)initWithId:(NSString *)uuid
-               withSession:(NSURLSession *)session
-              withDelegate:(EXDownloadDelegate *)delegate;
-{
-  if ((self = [super init])) {
-    _uuid = uuid;
-    _session = session;
-    _delegate = delegate;
-  }
-  return self;
-}
-
-@end
+typedef NS_ENUM(NSInteger, EXFileSystemHTTPMethod) {
+  EXFileSystemPostMethod = 0,
+  EXFileSystemPutMethod = 1,
+  EXFileSystemPatchMethod = 2,
+};
 
 @interface EXFileSystem ()
 
-@property (nonatomic, strong) NSMutableDictionary<NSString *, EXDownloadResumable*> *downloadObjects;
+@property (nonatomic, strong) NSMutableArray<NSURLSession *> *sessions;
```

Comment:
I personally would find the latter order of references more clear, it would also let us hide the session creation logic inside different `EXSessionTask` subclasses. Do you think it would make sense?
 
![excalidraw-202012311929-31](https://user-images.githubusercontent.com/1151041/78793970-68a56c80-79b3-11ea-8f97-50b373a14508.png)

Note: This is an opinion, so it may make sense to leave it as it is! Nonetheless, I think that providing such a simple overview of the structure would be profitable for future maintainers (I don't think the fact that `NSURLSession` keeps a strong reference to its `delegate` is a very common pattern, so the way we depend on it may be non-obvious for first-time readers).

Reply:
I partially agree with you. However, iOS by itself store the strong reference to the delegate (according to the: https://developer.apple.com/documentation/foundation/nsurlsession/1411597-sessionwithconfiguration?language=objc). We can't change it. And I've based my implementation on the previous one.

 I think we need to change how we use sessions. Currently, we use one session per task not just one per application(`experience`). This change would make this code more cleaner. 

I know that we discussed it earlier and I wasn't convinced to do it. But maybe it'll be better to do it in that way? What do you think?

Reply:
I've decided to try to structure this in a different way, which is similar to what you proposed. I think it is a much better right now.